### ISSUE TITLE
IMAGE CAPTIONING

### ISSUE THREAD

**Issue by jiya-coder8563** [jiya-coder8563/CodSoft-#3]
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Embedding, LSTM, Dense, Add, Input
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
import matplotlib.pyplot as plt
from PIL import Image

# Step 1: Load and Preprocess Image using VGG16
def preprocess_image(img_path):
    img = Image.open(img_path)
    img = img.resize((224, 224))
    img_array = np.array(img)
    img_array = np.expand_dims(img_array, axis=0)
    img_array = preprocess_input(img_array)  # Preprocess for VGG16
    return img_array

# Load pre-trained VGG16 model for feature extraction
vgg_model = VGG16(include_top=False, weights='imagenet', input_shape=(224, 224, 3))
vgg_model.trainable = False  # Freeze the VGG16 model
vgg_features = Model(inputs=vgg_model.input, outputs=vgg_model.output)

# Extract image features from VGG16
def extract_features(img_path):
    img_array = preprocess_image(img_path)
    features = vgg_features.predict(img_array)
    features = features.flatten()  # Flatten the feature map to a vector
    return features

# Step 2: Prepare the LSTM model (Decoder)
def define_captioning_model(vocab_size, max_length):
    # Image feature input
    img_input = Input(shape=(4096,))
    img_dense = Dense(256, activation='relu')(img_input)

    # Sequence input (captions as sequences of words)
    seq_input = Input(shape=(max_length,))
    seq_embed = Embedding(vocab_size, 256)(seq_input)
    seq_lstm = LSTM(256)(seq_embed)

    # Merge image features and LSTM output
    merged = Add()([img_dense, seq_lstm])
    output = Dense(vocab_size, activation='softmax')(merged)

    model = Model(inputs=[img_input, seq_input], outputs=output)
    model.compile(loss='categorical_crossentropy', optimizer='adam')
    return model

# Step 3: Generate Captions
def generate_caption(model, img_path, max_length, tokenizer):
    features = extract_features(img_path)
    sequence = [tokenizer.word_index['startseq']]  # Start the sequence with the start token
    caption = []
    
    # Generate words one at a time
    for _ in range(max_length):
        sequence_padded = pad_sequences([sequence], maxlen=max_length)
        prediction = model.predict([features, sequence_padded], verbose=0)
        predicted_word_index = np.argmax(prediction)
        predicted_word = tokenizer.index_word[predicted_word_index]
        
        if predicted_word == 'endseq':
            break
            
        caption.append(predicted_word)
        sequence.append(predicted_word_index)
    
    return ' '.join(caption)

# Step 4: Load and Preprocess Data (assuming you have tokenized captions)
# For simplicity, assume we have the 'tokenizer' and 'vocab_size' variables ready.
# tokenizer = ...
# vocab_size = ...
# max_length = ...

# Example usage:
# img_path = "path_to_image.jpg"
# caption = generate_caption(model, img_path, max_length, tokenizer)
# print("Generated Caption:", caption)
